wpilibj/src/main/java/edu/wpi/first/wpilibj/kinematics/DifferentialDriveOdometry.java

```diff
@@ -0,0 +1,128 @@
+    m_pose = pose;
+    m_prevAngle = pose.getRotation();
+  }
+
+  /**
+   * Updates the robot's position on the field using forward kinematics and
+   * integration of the pose over time.
+   *
+   * @param leftEncoder  The value of the left encoder (position). The units for x and y of
+   *                     the returned pose are the same as the units you pass in here. Therefore, it
+   *                     is advised that you convert the raw encoder value into meters, feet, or
+   *                     inches.
+   * @param rightEncoder The value of the right encoder (position). The units for x and y
+   *                     of the returned pose are the same as the units you pass in here. Therefore,
+   *                     it is advised that you convert the raw encoder value into meters, feet, or
+   *                     inches.
+   * @return The new pose of the robot.
+   */
+  public synchronized Pose2d update(double leftEncoder, double rightEncoder) {
+    final var deltaLeft = leftEncoder - m_prevLeftEncoder;
+    final var deltaRight = rightEncoder - m_prevRightEncoder;
+
+    final var chassisState = m_kinematics.toChassisSpeeds(
+        new DifferentialDriveWheelSpeeds(deltaLeft, deltaRight));
+
+    m_pose = m_pose.exp(new Twist2d(chassisState.dx, chassisState.dy, chassisState.dtheta));
+
+    m_prevLeftEncoder = leftEncoder;
+    m_prevRightEncoder = rightEncoder;
+    m_prevAngle = m_prevAngle.plus(new Rotation2d(chassisState.dtheta));
+
+    return m_pose;
+  }
+
+  /**
+   * Updates the robot's position on the field using forward kinematics and
+   * integration of the pose over time. This method uses input from the
+   * gyro instead of pure forward kinematics for angular data.
+   *
+   * @param leftEncoder  The value of the left encoder (position). The units for x and y of
+   *                     the returned pose are the same as the units you pass in here. Therefore, it
+   *                     is advised that you convert the raw encoder value into meters, feet, or
+   *                     inches.
+   * @param rightEncoder The value of the right encoder (position). The units for x and y
+   *                     of the returned pose are the same as the units you pass in here. Therefore,
+   *                     it is advised that you convert the raw encoder value into meters, feet, or
+   *                     inches.
+   * @param gyro         The yaw of the robot from a gyroscope.
+   * @return The new pose of the robot.
+   */
+  public synchronized Pose2d update(double leftEncoder, double rightEncoder, Rotation2d gyro) {
+    final var deltaLeft = leftEncoder - m_prevLeftEncoder;
+    final var deltaRight = rightEncoder - m_prevRightEncoder;
+
+    final var dx = (deltaLeft + deltaRight) / 2;
+    final var dtheta = gyro.minus(m_prevAngle);
+
+    m_pose = m_pose.exp(new Twist2d(dx, 0, dtheta.getRadians()));
+
+    m_prevLeftEncoder = leftEncoder;
+    m_prevRightEncoder = rightEncoder;
+    m_prevAngle = gyro;
```

Comment:
Seems like this is a great place to do some sensor fusion and combine encoders + gyro.  Thoughts?